pkg/security/security_profile/activity_tree/activity_tree.go

```diff
@@ -0,0 +1,381 @@
+		return true, nil
+	}
+	if err != nil {
+		return false, err
+	}
+	if node == nil {
+		// a process node couldn't be found or created for this event, ignore it
+		return false, err
+	}
+
+	// resolve fields
+	event.ResolveFieldsForAD()
+
+	// the count of processed events is the count of events that matched the activity dump selector = the events for
+	// which we successfully found a process activity node
+	at.Stats.processedCount[event.GetEventType()].Inc()
+
+	// insert the event based on its type
+	switch event.GetEventType() {
+	case model.ExecEventType:
+		// tag the matched rules if any
+		node.MatchedRules = model.AppendMatchedRule(node.MatchedRules, event.Rules)
+		return newProcessNode, nil
+	case model.FileOpenEventType:
+		return node.InsertFileEvent(&event.Open.File, event, generationType, at.Stats, at.shouldMergePaths, shadowInsertion)
+	case model.DNSEventType:
+		return node.InsertDNSEvent(event, generationType, at.Stats, at.DNSNames, shadowInsertion)
+	case model.BindEventType:
+		return node.InsertBindEvent(event, generationType, at.Stats, shadowInsertion)
+	case model.SyscallsEventType:
+		return node.InsertSyscalls(event, at.SyscallsMask)
+	}
+
+	return false, nil
+}
+
+// CreateProcessNode finds or a create a new process activity node in the activity dump if the entry
+// matches the activity dump selector.
+func (at *ActivityTree) CreateProcessNode(entry *model.ProcessCacheEntry, generationType NodeGenerationType, shadowInsertion bool) (node *ProcessNode, newProcessNode bool, err error) {
+	// drop processes with abnormal paths
+	if entry.GetPathResolutionError() != "" {
+		return nil, false, fmt.Errorf("process cache entry error: %s", entry.GetPathResolutionError())
+	}
+
+	// look for a ProcessActivityNode by process cookie
+	if entry.Cookie > 0 {
+		var found bool
+		node, found = at.CookieToProcessNode[entry.Cookie]
+		if found {
+			return node, false, nil
+		}
+	}
+
+	defer func() {
+		// if a node was found, and if the entry has a valid cookie, insert a cookie shortcut
+		if entry.Cookie > 0 && node != nil {
+			at.CookieToProcessNode[entry.Cookie] = node
+		}
+	}()
+
+	// find or create a ProcessActivityNode for the parent of the input ProcessCacheEntry. If the parent is a fork entry,
+	// jump immediately to the next ancestor.
+	parentNode, newProcessNode, err := at.CreateProcessNode(entry.GetNextAncestorBinary(), Snapshot, shadowInsertion)
+	if err != nil || newProcessNode && shadowInsertion {
+		return parentNode, newProcessNode, err
+	}
+
+	// if parentNode is nil, the parent of the current node is out of tree (either because the parent is null, or it
+	// doesn't match the dump tags).
+	if parentNode == nil {
+
+		// since the parent of the current entry wasn't inserted, we need to know if the current entry needs to be inserted.
+		if !at.validator.MatchesSelector(entry) {
+			return nil, false, nil
+		}
+
+		// go through the root nodes and check if one of them matches the input ProcessCacheEntry:
```

Comment:
I wonder if we should make this search before searching for parents.

Reply:
Happy to chat about it, but if the process in question is a root node, then this is almost what the algorithm will do first. More importantly I don't really think there is any statistical reason why a process generating events in a profile is most likely to be a root process, so I would actually stick to the generic solution as it would prevent going over the same nodes in the tree multiple times.

Reply:
Yep, here again I forgot to think about it with the cookies in mind. I was afraid to maybe do this check twice on the call path (for the node and its parent). But with cookies it should not happen, you're right!